WordPress/gutenberg
# (Github-Workflows)(DevEx) Wordpress Playground for Gutenberg PRs

## What?

Post a comment in PRs with a [WordPress Playground](https://wordpress.org/playground/) URL allowing devs to use it to test the changeset in the PR.

Inspired by the work done for Woocommerce by @samueljseay.

## Why?

To provide a quick way to test the changesets in a fully client-side-based disposable WordPress instance that can be brought up with a single click. No need to build the plugin locally and spin up `wp-env` if you don't want to :). Wordpress+Gutenberg manual testing bliss <3

## How?

The Playground is a novel way to quickly create disposable and fast WordPress instances in the browser. It also has supporting logic that allows its configuration to be orchestrated via the JS or URL params ([blueprints](https://wordpress.github.io/wordpress-playground/build-your-first-app)). We can pass a JSON in the URL, and the Playground app will use it as a definition for setting up the clientside WordPress instance with the plugins defined. In our case, we use the proxy provided by it, which exposes an interface to install plugins to the instance already.

❌  I tried different approaches (see the commit history), including one with a separate workflow that polled the `build-plugin-zip.yml` every x seconds. This worked and gave additional flexibility in the sense that we could add additional event criteria, like ignoring some files in the PR. See the implementation here: https://github.com/WordPress/gutenberg/pull/62393/commits/44badf80e4afdd1ac2fefaa06f37dcd9fc37dac3. However, it later became clear that it became a little too complex + it could cause throttling if there were a lot of PRs triggering it at the same time. it's a good example of how to use polling and the GH API to make a separate workflow depend on one another. Clunky but works.

✅  I ended up scrapping that and decided to implement it as a new job called `create-playground-link` inside the `build-plugin-zip.yml` workflow. It depends on the `build` job and only runs if the trigger event is `pull_request` and if `build` is successful. This means that the workflow won't try to run this job if it was started via the `workflow_dispatch` event, so it won't affect the release flow in any way. The job checks- out Gutenberg and runs a JS script that builds the JSON blueprint and encodes it in the playground URL. Finally, it posts the blueprint in a comment to the PR (or updates it if it already exists).

💡 I borrowed from the [work done in WooCommerce](https://github.com/woocommerce/woocommerce/blob/bbeaac4892e73866af9f60da05fe1a5838cc4bfb/.github/workflows/scripts/generate-playground-blueprint.js) and adapted it to the Gutenberg workflows. Kudos to @samueljseay for the initial work there! 🙇🏻 

## Testing Instructions

### Testing the workflow - opening a PR:

1. Fork my Gutenberg repo in https://github.com/fullofcaffeine/gutenberg;
1. Create a new branch from `trunk` and push it, create a new PR;
1. Go to the `Actions` tab and find the "Build Gutenberg Plugin Zip` action on the left sidebar, click on it and click on the pending one - you'll need to wait a bit here until an agent picks it up, it can take a while;
1. It should start running and skip quite a few jobs, running only the "Build Release Artifact" job (because it was started by a pull request event). Wait for it to finish (brew some coffee or tea, as it can take up to 10m).
1. Once it finishes, it should then run the "Create WordPress Playground Link" job which should run pretty quickly - and once this one finishes, the comment should then appear in your PR:

<img width="1117" alt="Screenshot 2024-06-06 at 8 12 29 p m" src="https://github.com/WordPress/gutenberg/assets/81248/97ac2ccc-e81a-4204-bc1a-ff1d3d9357c9">


<img width="883" alt="Screenshot 2024-06-06 at 8 14 09 p m" src="https://github.com/WordPress/gutenberg/assets/81248/43f43b57-e2d9-424a-a9a9-b44cadf54127">

### Editing a PR (pushing updates):

1. Push any changes to the PR;
2. You should not see any duplicate comments.


### Testing that the link works

After you have access to the link in the message from your test repo/PR, if you click it, you'll notice an error like this:

<img width="1512" alt="Screenshot 2024-06-06 at 8 22 40 p m" src="https://github.com/WordPress/gutenberg/assets/81248/0a0c79fa-7cbe-4508-8d91-0d42b84a49bd">

This is because the blueprint JSON refers to a PR that the playground proxy can't find. You'll need to edit the JSON in the URL to point to a PR that actually exists in the official Gutenberg repo, the repo that the proxy from playground.wordpress.net is set up to look into (unless you install playground somewhere or locally and proxy it!). This won't happen in production as all URLs will be generated from PRs in the official GB repo.

So, you can do the following:
1. Copy the link in your test PR (or use this PR: https://github.com/fullofcaffeine/gutenberg/pull/25);
3. Paste it in a text editor, find the "pr=x" part where x is the number of your PR;
4. Change it to an ID of any existing PR in Gutenberg, e.g `62391`;
5. Now paste it into a new tab in your browser, and it should work fine (see screenshot below). You'll notice the loading label ("Downloading Gutenberg PR xx) will still refer to the old Gutenberg number (see screencast below), but that shouldn't affect anything - if it makes you feel better, you can also edit the number there in the blueprint JSON :).

## Screenshots or screencast 

<video src="https://github.com/WordPress/gutenberg/assets/81248/c5782699-652f-42bc-a07a-d538515f9e30" controls="controls" style="max-width: 730px;"></video>
